### PR TITLE
annotate ERTP with Data/Far

### PR DESCRIPTION
**Pull request by warner**
This updates the ERTP and CapTP packages to use Data and Far (#2018) as necessary to allow tests to pass while the marshal.js "throw errors upon unmarked empty objects and unmarked pass-by-reference objects" switches are turned on.

Note that all vats must change their `buildRootObject` function to mark the return value with `Far`, as well as any other code that might run in a vat and wants to traffic in a pass-by-reference object. In general, any time you see `harden({ methods.. })`, you should replace it with `Far('interface name', { methods.. })`. And any time you see `harden({})` where the object is meant to be pass-by-copy, you should replace it with `Data({})`. When #2018 is all done, we'll revert the Data changes (so `harden({})` will be pass-by-copy), but `Far` is the new normal.

This builds upon the 2018-update-swingset-far branch (which does the same thing for swingset tests), and should not land until PR #2530 has landed.

@katelynsills please take a look at the ERTP changes to see if I correctly figured out which things are pass-by-copy `Data` and which things are pass-by-reference `Far`. Also see if the interface names I applied make sense. I think they're mostly cosmetic, but it'll probably save some confusion down the road if they aren't completely misleading.

@michaelfig same for CapTP please.

### CODE REVIEW

File: packages/ERTP/src/displayInfo.js
Comment (katelynsills) [inline packages/ERTP/src/displayInfo.js:62]: I remember @erights saying that we are adding a property that might show up in something like `Reflect.ownKeys`. Was that for Far objects only?

**Review by katelynsills — APPROVED** [review]
ERTP changes look correct to me. Thank you for doing this!

File: packages/ERTP/src/displayInfo.js
Comment (dtribble) [reply packages/ERTP/src/displayInfo.js:62]: That's temporarily also for Data objects. Until this conversion is complete, at which point it's only Far.

Comment (katelynsills) [reply packages/ERTP/src/displayInfo.js:62]: Ok, in that case, this check will break, I think.

**Comment by katelynsills** [comment]
I realized that the `displayInfo` code will break, and that I should also look for things that should have been changed, but were not, so I'm withdrawing my approval until I can take a closer look. Would it be ok if I make a PR on top for the necessary changes?

File: packages/ERTP/src/displayInfo.js
Comment (warner) [reply packages/ERTP/src/displayInfo.js:62]: BTW in doing the Zoe conversion, I'm experimenting with the following change to `cleanProposal.js`, which has a similar issue:

```js

const rejectSymbols = record => {
  // Reject all symbols except for 'passStyle', which is used by the Data()
  // marker. When we remove Data, this should revert back to rejecting
  // absolutely all symbols.
  for (const sym of Object.getOwnPropertySymbols(record)) {
    if (sym !== PASS_STYLE) {
      assert.fail(X`no symbol properties allowed: ${q(String(sym))}`);
    }
  }
};


// Assert that the keys of `record` are all in `allowedKeys`. If a key
// of `record` is not in `allowedKeys`, throw an error. If a key in
// `allowedKeys` is not a key of record, we do not throw an error.
const assertKeysAllowed = (allowedKeys, record) => {
  const keys = Object.getOwnPropertyNames(record);
  assertSubset(allowedKeys, keys);
  rejectSymbols(record);
};

// similar change to `cleanKeywords()`
```

(which also depends upon a patch to marshal.js to expose the `PASS_STYLE` symbol)

Comment (katelynsills) [reply packages/ERTP/src/displayInfo.js:62]: Actually based on discussion with @erights on keybase, I think this check is correct, as Far doesn't add extra properties to the object itself, but instead adds a new object in the inheritance chain.

**Comment by katelynsills** [comment]
I made a PR stacked on this, which changes `amounts` to use `Data`. https://github.com/Agoric/agoric-sdk/pull/2535

File: packages/ERTP/src/displayInfo.js
Comment (warner) [reply packages/ERTP/src/displayInfo.js:62]: We just decided to change the way Data is represented to avoid this, in #2538. I won't land this PR until that fix has landed and I've reviewed the use of `ownKeys` in ERTP.

**Comment by warner** [comment]
The changes in #2535 are correct but unnecessary: we only really need `Data` on objects that are empty (or which might be, e.g. if we build them with `{ ...otherObject }` and `otherObject` might be empty). Since we want to strip out `Data` when we're doing with the #2018 transition process, I think it might be good policy to *not* add `Data` everywhere it's logically correct, and restrict it to the ambiguous cases, so there's less to remove later.

If that matches your understanding too, I'll reject and close #2535.

**Comment by warner** [comment]
I pushed a new version that removes several not-strictly-necessary `Data()` markers as mentioned in #2535. It also rebases onto a newer trunk.

This should not land until #2538 is landed. We should review `displayInfo.js` and especially the prototype check in `coerceDisplayInfo` and see what still makes sense to assert.

File: packages/ERTP/src/displayInfo.js
Comment (warner) [inline packages/ERTP/src/displayInfo.js:63]: If the `allegedDisplayInfo` that gets passed in was marked as `Data`, this prototype check will now fail (`Data` is now marked with a special inserted prototype, just like `Remotable`/`Far`).

I'd like someone who understands the intent better than me to think about whether that's the correct behavior or not.

To distinguish Data from Far, if it really matters, it might be best to use `passStyleOf()`.

**Comment by warner** [comment]
Ok this is rebased onto a trunk that has #2538 landed, and I did a quick scan to see that it makes sense to me. Time for more experienced eyeballs to take a look: @katelynsills @michaelfig when you get a chance.  @erights you might also have thoughts about `coerceDisplayInfo`.

File: packages/ERTP/src/displayInfo.js
Comment (katelynsills) [reply packages/ERTP/src/displayInfo.js:63]: `allegedDisplayInfo` should ideally always be Data. In fact, all this code around lines 62-63 is trying to handle the degenerate case in which an empty object is wrongly treated as Far. If we think in this PR that we can prevent an empty object as being wrongly treated as Far, we can remove all of this code.

Comment (katelynsills) [reply packages/ERTP/src/displayInfo.js:63]: I believe these lines can be changed to:

```js
export const assertDisplayInfo = allegedDisplayInfo => {
  const displayInfoKeys = harden(['decimalPlaces']);
  assertKeysAllowed(displayInfoKeys, allegedDisplayInfo);
};

export const coerceDisplayInfo = allegedDisplayInfo => {
  if (allegedDisplayInfo === undefined) {
    return undefined;
  }
  assert(passStyleOf(allegedDisplayInfo) === 'copyRecord');
  allegedDisplayInfo = pureCopy(allegedDisplayInfo);
  assertDisplayInfo(allegedDisplayInfo);
  return allegedDisplayInfo;
};
```

Comment (warner) [reply packages/ERTP/src/displayInfo.js:63]: @erights should we export a RECORD_STYLE (`= 'copyRecord'`) to match the `REMOTE_STYLE` we currently export? We'd need something stable if we want to be able to ask `passStyleOf` about pass-by-copy instead of asking about something not being pass-by-reference.

Comment (warner) [reply packages/ERTP/src/displayInfo.js:63]: @katelynsills sounds like a good change, I've filed #2553 to make that cleanup after we land this (and after we decide on a RECORD_STYLE export).

Comment (erights) [reply packages/ERTP/src/displayInfo.js:63]: See my response to @dtribble at https://github.com/Agoric/agoric-sdk/pull/2546#discussion_r583402307 :

> `REMOTE_STYLE` is an outlier because we know the name it has (`'presence'`) is not the name we want (`'remote'`). For the other's, the string literal style is pervasive in our code. We should discuss sometime the tradeoffs. There are a lot of places in JS where a misspelling can lead to something silently going astray, such as property or method dispatch. Our properjs typescript/jsdoc types help a lot with these. They also do OR typing of an enumerated set of string literals. So I don't know that the string-literal-as-enum is really a different hazard than property or method names. 
>
> We should definitely examine this question. But I'm not gonna take on a pervasive style change in this narrowly targeted PR.

My position remains that we should continue to use literal strings, and enlist the typing system to help us discover when we're doing it wrong. But I'm sure this is not settled and something we need to argue through, experiment with the typings of, and settle on one consensus style. Attn @dtribble @michaelfig 

In the meantime, please do not introduce a `RECORD_STYLE`.

**Review by katelynsills — APPROVED** [review]
LGTM! Thanks!

**Comment by warner** [comment]
Since we have @katelynsills 's approval, I'm going to rebase and land the ERTP portion of this, and make a separate PR for the captp half so @michaelfig can have more time to review it.

**Pull request merged**
